apps/root/src/pages/strategy-guardian-detail/strategy-management/withdraw/hooks/useEarnWithdrawActions.tsx

```diff
@@ -0,0 +1,182 @@
+import React from 'react';
+import { useEarnManagementState } from '@state/earn-management/hooks';
+import { DisplayStrategy, EarnPermission, EarnWithdrawTypeData, TransactionTypes } from 'common-types';
+import { FormattedMessage } from 'react-intl';
+import { parseUnits } from 'viem';
+import useTrackEvent from '@hooks/useTrackEvent';
+import useActiveWallet from '@hooks/useActiveWallet';
+import { shouldTrackError } from '@common/utils/errors';
+import useErrorService from '@hooks/useErrorService';
+import { Typography } from 'ui-library';
+import useTransactionModal from '@hooks/useTransactionModal';
+import { useTransactionAdder } from '@state/transactions/hooks';
+import useEarnService from '@hooks/earn/useEarnService';
+import { getProtocolToken, getWrappedProtocolToken } from '@common/mocks/tokens';
+import { isSameToken } from '@common/utils/currency';
+import useSupportsSigning from '@hooks/useSupportsSigning';
+
+interface UseEarnWithdrawActionsParams {
+  strategy?: DisplayStrategy;
+}
+
+const useEarnWithdrawActions = ({ strategy }: UseEarnWithdrawActionsParams) => {
+  const asset = strategy?.asset;
+  const activeWallet = useActiveWallet();
+  const trackEvent = useTrackEvent();
+  const errorService = useErrorService();
+  const earnService = useEarnService();
+  const [currentTransaction, setCurrentTransaction] = React.useState('');
+  const { withdrawAmount: assetAmountInUnits, withdrawRewards } = useEarnManagementState();
+  const addTransaction = useTransactionAdder();
+  const [shouldShowConfirmation, setShouldShowConfirmation] = React.useState(false);
+  const [, setModalLoading, setModalError, setModalClosed] = useTransactionModal();
+  const hasSignSupport = useSupportsSigning();
+
+  const onWithdraw = React.useCallback(async () => {
+    const notWithdrawing = !assetAmountInUnits && !withdrawRewards;
+    if (!asset || !activeWallet?.address || !strategy || notWithdrawing) return;
+
+    const currentPosition = strategy.userPositions?.find((position) => position.owner === activeWallet.address);
+    if (!currentPosition) return;
+
+    const hasPermission = await earnService.companionHasPermission(currentPosition.id, EarnPermission.WITHDRAW);
+
+    const protocolToken = getProtocolToken(strategy.farm.chainId);
+    const wrappedProtocolToken = getWrappedProtocolToken(strategy.farm.chainId);
+
+    // Protocol tokens will be unwrapped
+    const assetIsWrappedProtocol = isSameToken(wrappedProtocolToken, asset);
+
+    try {
+      setModalLoading({
+        content: (
+          <>
+            <Typography variant="bodyRegular">
+              <FormattedMessage
+                description="earn.strategy-management.withdraw.modal.loading"
+                defaultMessage="Withdrawing funds from {farm}"
+                values={{ farm: strategy.farm.name }}
+              />
+            </Typography>
+            {assetIsWrappedProtocol && !hasPermission && hasSignSupport && (
```

Comment:
I know I've said otherwise before, but should we add a step actually to sign the permit?